// entity/context.js
/*
 *    Date: 2018 11 22
 * Project: Nodas - EntityManager
 *
 * Copyright 2018 SIT srl
 */

var entTemp = require(process.cwd()+"/../lib/entityTemplate").entityTemplate;
var async = require("async");
var util = require("util");
var self = {};

/*
 * Context-Table array
 */
var aCtxTable = [
  ["operator","ctx_operator"],
  ["wgServer","wg_server"],
  ["wgGeometry","wg_geometry"],
  ["wgLayerType","wg_layer_type"],
  ["wgStyleType","wg_style_type"],
  ["wgStyleFill","wg_style_fill","descr"],
  ["wgStyleShape","wg_style_shape","descr"],
  ["wgStyleStroke","wg_style_stroke","descr"],
  ["localita","grafo.view_localita"],
  ["municipio","grafo.view_municipio"],
  ["tipoLimAmm","grafo.tipo_lim_amm"],
  ["nodoTipo","grafo.nodo_tipo"],
  ["civicoLatoStrada","grafo.civico_lato_strada"]
];

/*
 * Table-ContextUrl object
 */
var objTableCtxUrl = {
  //"grafo.arco": "/arco/master",
  "grafo.arco_uso": "/arcoDizionari/master?dict=arco_uso",
  "grafo.arco_tipo": "/arcoDizionari/master?dict=arco_tipo",
  "grafo.arco_sede": "/arcoDizionari/master?dict=arco_sede",
  "grafo.arco_fondo": "/arcoDizionari/master?dict=arco_fondo",
  "grafo.arco_fonte": "/arcoDizionari/master?dict=arco_fonte",
  "grafo.arco_classe": "/arcoDizionari/master?dict=arco_classe",
  "grafo.arco_marcia": "/arcoDizionari/master?dict=arco_marcia",
  "grafo.arco_livello": "/arcoDizionari/master?dict=arco_livello",
  "grafo.arco_origine": "/arcoDizionari/master?dict=arco_origine",
  "grafo.arco_portata": "/arcoDizionari/master?dict=arco_portata",
  "grafo.arco_sezione": "/arcoDizionari/master?dict=arco_sezione",
  "grafo.arco_proprieta": "/arcoDizionari/master?dict=arco_proprieta",
  "grafo.arco_tipologia": "/arcoDizionari/master?dict=arco_tipologia",
  "grafo.arco_viabilita": "/arcoDizionari/master?dict=arco_viabilita",
  "grafo.arco_strada_cs": "/arcoDizionari/master?dict=arco_strada_cs",
  "grafo.arco_stato_cons": "/arcoDizionari/master?dict=arco_stato_cons",
  "grafo.arco_class_funz": "/arcoDizionari/master?dict=arco_class_funz",
  "grafo.arco_fondazione": "/arcoDizionari/master?dict=arco_fondazione",
  "grafo.arco_carreggiata": "/arcoDizionari/master?dict=arco_carreggiata",
  "grafo.arco_funzionalita": "/arcoDizionari/master?dict=arco_funzionalita",
  "grafo.arco_pavimentazione": "/arcoDizionari/master?dict=arco_pavimentazione",
  "grafo.arco_stato_esercizio": "/arcoDizionari/master?dict=arco_stato_esercizio",
  "grafo.arco_senso_percorrenza": "/arcoDizionari/master?dict=arco_senso_percorrenza",
  "grafo.arco_mot_ridenominazione": "/arcoDizionari/master?dict=arco_mot_ridenominazione",
  //"grafo.civico": "/civico/master",
  "grafo.civico_mot_cessazione": "/civicoMotCessazione/master",
  "grafo.civico_mot_inserimento": "/civicoMotInserimento/master",
  "grafo.civico_tipo_ingresso": "/civicoTipoIngresso/master",
  //"grafo.edificio": "/edificio/master",
  "grafo.edificio_diff_catasto": "/edificioDiffCatasto/master",
  "grafo.edificio_mot_cessazione": "/edificioMotCessazione/master",
  "grafo.edificio_stato": "/edificioStato/master",
  "grafo.edificio_tipo": "/edificioTipo/master",
  "grafo.edificio_uso_prevalente": "/edificioUsoPrevalente/master",
  //"grafo.nodo": "/nodo/master",
  //"grafo.via": "/via/master",
  "grafo.via_classificazione": "/viaClassificazione/master",
  "grafo.via_mot_cessazione": "/viaMotCessazione/master",
  "grafo.via_tipo": "/viaTipo/master",
  "grafo.via_tipo_numero": "/viaTipoNumero/master"
};

/*
 * Context entity
 */
var context = function()
{
  context.super_.call(this);

  this.moduleName = "Entity context";
  this.entityName = "context";

  this.ctxObj = null;
}

/*
 * Inheritance
 */
util.inherits(context,entTemp);

context.prototype.init = function(options)
{
  self = this;
  context.super_.prototype.init.call(self,options);
}

context.prototype.master = function(params,callback)
{
  if (self.ctxObj)
  {
    callback(null,self.ctxObj);
    return;
  }

  // Build context
  var queryArray = [];
  self.ctxObj = {};

  for (var i = 0;i < aCtxTable.length;i++)
  {
    queryArray.push({
      qOpt:
      {
        queryName: aCtxTable[i][0],
        fields: self.crudUtils.ALL_FIELDS,
        from: [{name:aCtxTable[i][1], type:self.crudUtils.TABLE}],
        ordering: [{
          field: aCtxTable[i][2] || "name",
          orderType: self.crudUtils.ASC
        }]
      },
      qVal: []
    });
  }

  async.each(queryArray,loadTable,function(err)
  {
    callback(err,self.ctxObj);
  });
}

/*
 * New methods
 */
context.prototype.tableInfo = function(opt,callback)
{
  if (!opt.object.table)
  {
    callback({message: "Missing table parameter"});
    return;
  }

  /*
   * Get table info from db
   */
  var qOpt = {queryString: `SELECT
    c.column_name AS column,
    c.data_type AS type,
    kcu.constraint_name,
    tc.constraint_type,
    ccu.table_schema AS fk_schema,
    ccu.table_name AS fk_table,
    ccu.column_name AS fk_column
    FROM information_schema.columns c
    LEFT JOIN information_schema.key_column_usage kcu ON
      c.table_name = kcu.table_name AND c.column_name = kcu.column_name
    LEFT JOIN information_schema.table_constraints tc ON
      tc.constraint_name = kcu.constraint_name
    LEFT JOIN information_schema.constraint_column_usage ccu ON
      ccu.constraint_name = kcu.constraint_name
    WHERE c.table_name = $1`
  };

  self.crud.select(qOpt,[{value: opt.object.table}],function(err,res)
  {
    if (err)
    {
      callback(err,null);
      return;
    }

    /*
     * Process response
     */
    var aRet = [];

    for (var j = 0;j < res.result.length;j++)
    {
      var inObj = res.result[j];
      var outObj = {
        column: inObj.column,
        type: inObj.type,
        context: null,
        url: null,
        key: null
      };

      if (inObj.constraint_type == "FOREIGN KEY")
      {
        outObj.context = ctxNameForTable(inObj.fk_schema,inObj.fk_table);
        outObj.url = objTableCtxUrl[`${inObj.fk_schema}.${inObj.fk_table}`] || null;
        outObj.key = inObj.fk_column;
      }

      aRet.push(outObj);
    }

    callback(null,{result: aRet});
  });
}

/*
 * Exports
 */
exports.context = context;

/*
 * Private functions
 */
function loadTable(obj,callback)
{
  self.crud.select(obj.qOpt,obj.qVal,function(err,res)
  {
    if (!err && res && res.result)
      self.ctxObj[obj.qOpt.queryName] = res.result;

    callback(err);
  });
}

function ctxNameForTable(schema,table)
{
  if (schema && schema != "public")
    table = `${schema}.${table}`;

  for (var j = 0;j < aCtxTable.length;j++)
    if (aCtxTable[j][1] == table)
      return aCtxTable[j][0];

  return null;
}
